controllers/gatekeepersync/health.go

```diff
@@ -0,0 +1,95 @@
+// Copyright Contributors to the Open Cluster Management project
+
+package gatekeepersync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apiWatch "k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/watch"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+
+	"open-cluster-management.io/governance-policy-framework-addon/controllers/utils"
+)
+
+// GatekeeperInstallationChecker is a health checker for a health endpoint that fails if Gatekeeper's installation
+// status changes or the passed in health checker functions fail. This is useful for Kubernetes to trigger a restart
+// to either enable or disable the gatekeeper-constraint-status-sync controller based on the Gatekeeper installation
+// status.
+func GatekeeperInstallationChecker(
+	ctx context.Context, dynamicClient dynamic.Interface, checkers ...healthz.Checker,
+) (
+	healthz.Checker, bool, error,
+) {
+	fieldSelector := fmt.Sprintf("metadata.name=constrainttemplates.%s", utils.GvkConstraintTemplate.Group)
+	timeout := int64(30)
+	crdGVR := schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1",
+		Resource: "customresourcedefinitions",
+	}
+
+	listResult, err := dynamicClient.Resource(crdGVR).List(
```

Comment:
If this is the first thing done by the dynamic client (and I think it might be) then I think it could take a long time to do its API discovery... you might already know more about that than me, so correct me if we don't need to be concerned with that here. But if this `List` times out (because of that issue), then the error would bubble up to `main`, and we'd effectively restart the container. But restarting the container will just cause it to have the same problem again, unless I'm missing something.

Reply:
The dynamic client doesn't perform a discovery from what I can tell in the code. It just uses the GVR to construct the Kubernetes API endpoint URL.